### ISSUE TITLE
assert_email_delivered_with unintuitive test syntax

### ISSUE THREAD

**Issue by blacksph3re** [thoughtbot/bamboo#444]
To test on certain fields it requires a weird pattern matching, as the passed keywords are matched directly against the mail structure. This makes regex matching for these fields impossible, as the field of the mail structure is passed directly to regex.match

Concretely it is:

- from: `from: {nil: "from@address.com"}`
- html_body: `html_body: {:ok, "<some>html</some>"}`
- to: `to: [nil: "to@address.com"]`

I would find it more intuitive if I wouldn't have to insert those maps and lists, also regex matching would work then. If I try with the intuitive way (`assert_email_delivered_with(to: "to@address.com")`) I get the error:

```
The parameters given do not match.
     
       Parameters given:
     
         [to: "to@address.com"]
     
       Email recieved:
     
         %{assigns: :assigns_removed_for_testing, attachments: [], bcc: [], cc: [], from: {nil, "from@address.com"}, headers: %{}, html_body: {:ok, "<!-- some html-->"}, private: %{}, subject: "some subject", text_body: nil, to: [nil: "to@address.com"]}

```

**Comment by lancejjohnson** [comment]
@blacksph3re Thanks for submitting this issue recommending a possible improvement. Would you be willing to provide an example test that uses the syntax as it stands followed by the same test with the syntax that you're recommending? I think that will help clarify the request.

**Comment by tcoopman** [comment]
Some examples:

```elixir
# instead of
assert_email_delivered_with(from: {nil, "foo@bar.com"})
assert_email_delivered_with(to: [nil, "foo@bar.com"])
# be able to use
assert_email_delivered_with(from: "foo@bar.com")
assert_email_delivered_with(to: "foo@bar.com")
```

**Comment by germsvel** [comment]
It seems like this was resolved by https://github.com/thoughtbot/bamboo/pull/482, so I'm going to close this. 

If this is still an issue, please let me know. Happy to reopen it.

**Comment by randycoulman** [comment]
I realize this issue has been closed for a while, but I just upgraded to v2.2.0 and checked to see if this was fixed for us. It turns out it isn't fixed by #482.

I investigated, and it looks like `Bamboo.Formatter` isn't having the desired effect:

```elixir
my_app (1)> Bamboo.Formatter.format_email_address([nil: "someone@example.com"], %{type: :to})
[nil: "someone@example.com"]
my_app (2)> Bamboo.Formatter.format_email_address({nil, "someone@example.com"}, %{type: :from})
{nil, "someone@example.com"}
```

#482 added a test for this new behavior, so I'm not sure why it isn't working.  But looking at the various implementations of the Formatter protocol, it makes sense.

In the first case, there's a keyword list, which is the same as `[{nil, "someone@example.com"}]`. The implementation maps over the list, formatting each element.

That hits the implementation for a tuple (as does the second case above), which treats the tuple as an already-formatted email address and returns it unchanged.
